File: 02. CLIENTE WEB/Cliente_Web/controller/amortizacionController.js
const REST_ENDPOINT_CLIENTE = "https://localhost:7268/api/Cliente";
const REST_ENDPOINT_AMORTIZACION = "https://localhost:7268/api/Amortizacion";


// Referencias a elementos del DOM
const amortizacionForm = document.getElementById("amortizacionForm");
const cedulaClienteInput = document.getElementById("cedulaCliente");
const amortizacionContainer = document.getElementById("amortizacionContainer");
const amortizacionTableBody = document.getElementById("amortizacionTable").querySelector("tbody");

// Función genérica para enviar solicitudes REST
async function sendRestRequest(url, method = "GET", body = null) {
    console.log(`[REST REQUEST] ${method} ${url}`);
    if (body) console.log("[REST REQUEST] Body:", JSON.stringify(body));

    try {
        const options = {
            method,
            headers: { "Content-Type": "application/json" },
        };

        if (body) options.body = JSON.stringify(body);

        const response = await fetch(url, options);
        console.log(`[REST RESPONSE] Status: ${response.status}`);

        if (!response.ok) throw new Error(`HTTP error! Status: ${response.status}`);

        return await response.json();
    } catch (error) {
        console.error("[REST ERROR] Error en la solicitud:", error);
        throw error;
    }
}

// Obtener código de crédito basado en la cédula del cliente
async function obtenerCodigoCredito(cedula) {
    const url = `${REST_ENDPOINT_CLIENTE}/obtenerCodigoCliente/${cedula}`;
    return await sendRestRequest(url);
}

// Obtener tabla de amortización basada en el código de crédito
async function obtenerTablaAmortizacion(codCredito) {
    const url = `${REST_ENDPOINT_AMORTIZACION}/credito/${codCredito}`;
    return await sendRestRequest(url);
}

// Manejo del formulario
amortizacionForm.addEventListener("submit", async (event) => {
    event.preventDefault();
    const cedula = cedulaClienteInput.value.trim();

    if (!cedula) {
        alert("Ingrese una cédula válida.");
        return;
    }

    try {
        // Obtener código de crédito
        const codCredito = await obtenerCodigoCredito(cedula);
        if (!codCredito) {
            alert("No se encontró crédito asociado a esta cédula.");
            amortizacionContainer.style.display = "none";
            return;
        }

        // Obtener y mostrar tabla de amortización
        const amortizaciones = await obtenerTablaAmortizacion(codCredito);
        amortizacionTableBody.innerHTML = "";

        if (amortizaciones.length === 0) {
            alert("No se encontraron registros de amortización.");
            amortizacionContainer.style.display = "none";
            return;
        }

        amortizaciones.forEach(amortizacion => {
            const row = document.createElement("tr");
            row.innerHTML = `
                <td>${amortizacion.codAmortizacion}</td>
                <td>${amortizacion.codCredito}</td>
                <td>$${amortizacion.capitalPagado.toFixed(2)}</td>
                <td>$${amortizacion.interesPagado.toFixed(2)}</td>
                <td>${amortizacion.numCuota}</td>
                <td>$${amortizacion.saldo.toFixed(2)}</td>
                <td>$${amortizacion.valorCuota.toFixed(2)}</td>
                <td>${amortizacion.fechaPago}</td>`;
            amortizacionTableBody.appendChild(row);
        });

        amortizacionContainer.style.display = "block";
    } catch (error) {
        console.error("[AMORTIZACION] Error:", error);
        alert("Error al consultar la tabla de amortización.");
    }
});

File: 03. CLIENTE DE CONSOLA/CLIENTE_CONSOLA/CLIENTE_CONSOLA/ec.edu.monster.controller/AmortizacionController.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using ec.edu.monster.service;
using ec.edu.monster.model;

namespace ec.edu.monster.controller
{
    public class AmortizacionController
    {
        private readonly ApiService _apiService;

        public AmortizacionController()
        {
            _apiService = new ApiService();
        }

        // Consultar amortización utilizando la cédula para obtener el código del crédito
        public async Task ConsultarAmortizacionesPorCedula()
        {
            Console.Write("\nIngrese la cédula del cliente: ");
            string cedula = Console.ReadLine()?.Trim();

            if (string.IsNullOrEmpty(cedula))
            {
                Console.WriteLine("La cédula es obligatoria.");
                return;
            }

            // Obtener el código del cliente asociado a la cédula
            int codCliente = await _apiService.ObtenerCodigoCliente(cedula);

            if (codCliente == -1)
            {
                Console.WriteLine("No se encontró un cliente con la cédula ingresada.");
                return;
            }

            // Obtener el código del último crédito asociado al cliente
            int codCredito = await _apiService.ObtenerUltimoCredito(codCliente);

            if (codCredito == -1)
            {
                Console.WriteLine("No se encontró un crédito activo para el cliente.");
                return;
            }

            // Consultar la tabla de amortización usando el código del crédito
            List<Amortizacion> amortizaciones = await _apiService.ObtenerAmortizaciones(codCredito);

            if (amortizaciones == null || amortizaciones.Count == 0)
            {
                Console.WriteLine("No se encontraron amortizaciones para el crédito.");
                return;
            }

            Console.WriteLine("\n=== Tabla de Amortización ===");
            Console.WriteLine($"{"Cuota",-6} {"Pago",-10} {"Interés",-10} {"Capital Pagado",-15} {"Saldo",-10}");
            foreach (var amortizacion in amortizaciones)
            {
                Console.WriteLine($"{amortizacion.NumCuota,-6} ${amortizacion.ValorCuota,-10:F2} ${amortizacion.InteresPagado,-10:F2} ${amortizacion.CapitalPagado,-15:F2} ${amortizacion.Saldo,-10:F2}");
            }
        }
    }
}
